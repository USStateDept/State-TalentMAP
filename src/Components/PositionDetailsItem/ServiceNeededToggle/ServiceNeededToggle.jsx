import React, { Component } from 'react';
import FontAwesome from 'react-fontawesome';
import PropTypes from 'prop-types';

import InteractiveElement from '../../InteractiveElement';

import {
  EMPTY_FUNCTION,
  POSITION_DETAILS,
  USER_PROFILE,
} from '../../../Constants/PropTypes';

export default class ServiceNeededToggle extends Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }

  onClick() {
    const { position, onChange } = this.props;
    onChange(position.id, !this.checked);
  }

  get checked() {
    const { position } = this.props;
    return (position.is_highlighted || false);
  }

  get text() {
    return `${this.checked ? 'Unmark' : 'Mark'} as Service Need`;
  }

  get icon() {
    return this.checked ? 'check-square-o' : 'square-o';
  }

  render() {
    const props = this.props;
    const user = props.userProfile;
    const wrapper = {
      className: 'service-needed-toggle',
    };

    const isSuperuser = user.is_superuser;
    const options = {
      className: ['usa-button', props.className],
      onClick: this.onClick,
    };

    options.className = options.className
      .join(' ')
      .trim();

    return isSuperuser ?
      (<div {...wrapper}>
        <InteractiveElement {...options}>
          {!props.loading ?
            (<FontAwesome name={this.icon} />) :
            (<div className="ds-c-spinner" />)
          } {this.text}
        </InteractiveElement>
      </div>) : (null);
  }
}

ServiceNeededToggle.propTypes = {
  as: PropTypes.string.isRequired,
  className: PropTypes.string,
  position: POSITION_DETAILS,
  userProfile: USER_PROFILE,
  loading: PropTypes.bool.isRequired,
  onChange: PropTypes.func,
};

ServiceNeededToggle.defaultProps = {
  as: 'div',
  className: '',
  position: {},
  userProfile: {},
  loading: false,
  onChange: EMPTY_FUNCTION,
};
