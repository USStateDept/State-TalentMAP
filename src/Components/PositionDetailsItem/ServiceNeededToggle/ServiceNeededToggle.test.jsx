import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import ServiceNeededToggle from './ServiceNeededToggle';

import mock from '../../../__mocks__/detailsObject';

const userMock = {
  is_superuser: true,
};

const props = {
  position: mock,
  userProfile: userMock,
};

describe('ServiceNeededToggle', () => {
  beforeEach(() => {
    props.userProfile.is_superuser = true;
  });

  it('is defined', () => {
    const toggle = shallow(<ServiceNeededToggle {...props} />);
    expect(toggle.html()).not.toBeNull();
  });

  it('only renders when user is superuser', () => {
    props.userProfile.is_superuser = false;
    const toggle = shallow(<ServiceNeededToggle {...props} />);
    expect(toggle.html()).toBeNull();
  });

  it('verifies loading state renders and refreshes', () => {
    let toggle;

    toggle = shallow(<ServiceNeededToggle {...props} loading />);
    expect(toggle.find('.ds-c-spinner').length).toBe(1);

    toggle = shallow(<ServiceNeededToggle {...props} loading={false} />);
    expect(toggle.find('.ds-c-spinner').length).toBe(0);
  });

  it('matches snapshot', () => {
    const toggle = shallow(<ServiceNeededToggle {...props} />);
    expect(toJSON(toggle)).toMatchSnapshot();
  });
});
