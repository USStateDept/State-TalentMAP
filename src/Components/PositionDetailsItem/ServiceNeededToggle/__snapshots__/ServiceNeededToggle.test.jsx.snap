// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceNeededToggle matches snapshot 1`] = `
<div
  className="service-needed-toggle"
>
  <InteractiveElement
    className="usa-button"
    onClick={[Function]}
    type="div"
  >
    <FontAwesome
      name="square-o"
    />
     
    Mark as Service Need
  </InteractiveElement>
</div>
`;
